test/py/test_device_get_region_io_fds.py

```diff
@@ -0,0 +1,181 @@
+
+    newfds, ret = msg_fd(ctx, sock, VFIO_USER_DEVICE_GET_REGION_IO_FDS, payload, expect=0)
+
+    reply, ret = region_io_fds_reply.pop_from_buffer(ret)
+    ioevent, ret = sub_region_ioeventfd.pop_from_buffer(ret)
+
+    print(newfds)
+    for i in range(0,4):
+        os.write(newfds[i], c.c_ulonglong(10))
+        out = os.read(newfds[i], 8)
+        [out] = struct.unpack("@Q",out)
+        assert out  == 10
+
+def test_device_get_region_io_fds_full():
+
+    payload = region_io_fds_request(argsz = 16+(40*6), flags = 0, index = VFU_PCI_DEV_BAR2_REGION_IDX, count = 0)
+
+    newfds, ret = msg_fd(ctx, sock, VFIO_USER_DEVICE_GET_REGION_IO_FDS, payload, expect=0)
+
+    reply, ret = region_io_fds_reply.pop_from_buffer(ret)
+    ioevents = []
+    for i in range(0, reply.count):
+        ioevent, ret = sub_region_ioeventfd.pop_from_buffer(ret)
+        ioevents.append(ioevent)
+        os.write(newfds[ioevent.fd_index], c.c_ulonglong(1))
+
+    print(newfds, [i.fd_index for i in ioevents])
+
+    for i in range(0, reply.count):
+        out = os.read(newfds[ioevents[i].fd_index], ioevent.size)
+        [out] = struct.unpack("@Q",out)
+        assert out  == 1
+        assert ioevents[i].size == 8
+        assert ioevents[i].offset == 40 - (8 * i)
+
+def test_device_get_region_io_fds_fds_read_write_nothing():
+
+    payload = region_io_fds_request(argsz = 16, flags = 0, index = VFU_PCI_DEV_BAR2_REGION_IDX, count = 0)
+
+    newfds, ret = msg_fd(ctx, sock, VFIO_USER_DEVICE_GET_REGION_IO_FDS, payload, expect=0)
+
+    reply, ret = region_io_fds_reply.pop_from_buffer(ret)
+    assert reply.argsz == 16
+
+def test_device_get_region_io_fds_fds_read_write_dupe_fd():
+
+    t = eventfd(0,0)
+    assert vfu_create_ioeventfd(ctx, 6 * 8, 8, t, 0, VFU_PCI_DEV_BAR2_REGION_IDX) != -1
+    assert vfu_create_ioeventfd(ctx, 7 * 8, 8, t, 0, VFU_PCI_DEV_BAR2_REGION_IDX) != -1
+
+
+    payload = region_io_fds_request(argsz = 16+(40*8), flags = 0, index = VFU_PCI_DEV_BAR2_REGION_IDX, count = 0)
+
+    newfds, ret = msg_fd(ctx, sock, VFIO_USER_DEVICE_GET_REGION_IO_FDS, payload, expect=0)
+    reply, ret = region_io_fds_reply.pop_from_buffer(ret)
+
+    assert reply.count == 8
+    assert reply.argsz == 16+(40*8)
+
+    ioevents = []
+    for i in range(0, reply.count):
+        ioevent, ret = sub_region_ioeventfd.pop_from_buffer(ret)
+        ioevents.append(ioevent)
+
+    for i in range(2, 8):
+        os.write(newfds[ioevents[i].fd_index], c.c_ulonglong(1))
+
+    print(newfds, [i.fd_index for i in ioevents])
+
+    for i in range(2, 8):
+        out = os.read(newfds[ioevents[i].fd_index], ioevent.size)
+        [out] = struct.unpack("@Q",out)
+        assert out  == 1
+        assert ioevents[i].size == 8
+        assert ioevents[i].offset == 56 - (8 * i)
+
+
+    assert ioevents[0].fd_index == ioevents[1].fd_index
+    assert ioevents[0].offset != ioevents[1].offset
+
+    os.write(newfds[ioevents[0].fd_index], c.c_ulonglong(1))
```

Comment:
I think you want to write to the ioevent[1] fd_index too, then confirm that it reads on the server side as 2 ?